shelley/chain-and-ledger/executable-spec/src/BlockChain.hs

```diff
@@ -244,6 +253,134 @@ data Block crypto
     (TxSeq crypto)
   deriving (Show, Eq)
 
+-- The SegWits data type is only used for serializing and deserializing
+-- blocks using a "segregated witness" format:
+--
+--   block =
+--     [ header
+--     , transaction_bodies         : [* transaction_body]
+--     , transaction_witness_sets   : [* transaction_witness_set]
+--     ]
+--
+--     and transaction_witness_set is serialized depending on what
+--     values are present:
+--
+--  transaction_witness_set =
+--    (  0, vkeywitness
+--    // 1, $script
+--    // 2, [* vkeywitness]
+--    // 3, [* $script]
+--    // 4, [* vkeywitness],[* $script]
+--    )
+data SegWits crypto
+  = SegWitsVKey (WitVKey crypto)
+  | SegWitsVKeys (Set (WitVKey crypto))
+  | SegWitsScript (ScriptHash crypto) (MultiSig crypto)
+  | SegWitsScripts (Map (ScriptHash crypto) (MultiSig crypto))
+  | SegWitsAll (Set (WitVKey crypto)) (Map (ScriptHash crypto) (MultiSig crypto))
+  deriving (Show, Eq)
+
+segWitToTx
+  :: (Crypto crypto)
+  => (TxBody crypto, SegWits crypto)
+  -> Tx crypto
+segWitToTx (body, SegWitsVKey w) = Tx body (Set.singleton w) mempty
+segWitToTx (body, SegWitsVKeys ws) = Tx body ws mempty
+segWitToTx (body, SegWitsScript s m) = Tx body mempty (Map.singleton s m)
+segWitToTx (body, SegWitsScripts ss) = Tx body mempty ss
+segWitToTx (body, SegWitsAll ws ss) = Tx body ws ss
+
+txToSegWit
+  :: (Crypto crypto)
+  => Tx crypto
+  -> SegWits crypto
+txToSegWit (Tx _ ws ss) =
+  case (Set.toList ws, Map.toList ss) of
+    ([x]      , [])        -> SegWitsVKey x
+    (zs@(_:_), [])        -> SegWitsVKeys $ Set.fromList zs
+    ([]       , [p])       -> uncurry SegWitsScript p
+    ([]       , qs@(_:_)) -> SegWitsScripts $ Map.fromList qs
+    (zs       , qs)        -> SegWitsAll (Set.fromList zs) (Map.fromList qs)
+
+instance Crypto crypto
+  => ToCBOR (SegWits crypto)
+ where
+  toCBOR = \case
+    SegWitsVKey w ->
+      encodeListLen 2
+        <> toCBOR (0 :: Word8)
+        <> toCBOR w
+
+    SegWitsVKeys ws ->
+      encodeListLen 2
+        <> toCBOR (1 :: Word8)
+        <> toCBOR ws
+
+    SegWitsScript s m ->
+      encodeListLen 3
+        <> toCBOR (2 :: Word8)
+        <> toCBOR s
+        <> toCBOR m
+
+    SegWitsScripts ss ->
+      encodeListLen 2
+        <> toCBOR (3 :: Word8)
+        <> toCBOR ss
+
+    SegWitsAll ws ss ->
+      encodeListLen 3
+        <> toCBOR (4 :: Word8)
+        <> toCBOR ws
+        <> toCBOR ss
+
+instance Crypto crypto
+  => FromCBOR (SegWits crypto)
+ where
+  fromCBOR = decodeListLen >> decodeWord >>= \case
```

Comment:
We should compare the value from `decodeListLen` to the size of the record we get here

Reply:
yes indeed, thanks!